File: src/jarvis/tools/web.py
"""Web automation and query tools for Jarvis."""

import asyncio
import re
import shutil
import urllib.parse
import urllib.request
from typing import Optional, Tuple
from jarvis.core.logger import log


class WebNavigator:
    """Handles web browsing, searches, and direct YouTube video resolution."""

    def __init__(self):
        self._browser_bin = shutil.which("omarchy-launch-browser") or shutil.which("xdg-open") or "xdg-open"

    async def open_url(self, url: str) -> str:
        """Open a specific URL in the user's default browser."""
        url = url.strip()
        if not url.startswith(("http://", "https://")):
            url = f"https://{url}"

        log.info(f"[cyan]Opening URL in browser: {url}[/cyan]")
        try:
            proc = await asyncio.create_subprocess_exec(
                self._browser_bin,
                url,
                stdout=asyncio.subprocess.DEVNULL,
                stderr=asyncio.subprocess.DEVNULL,
                start_new_session=True,
            )
            try:
                await asyncio.wait_for(proc.wait(), timeout=0.15)
            except asyncio.TimeoutError:
                pass
            return f"Opened {url} in browser."
        except Exception as e:
            log.error(f"Failed to open URL '{url}': {e}")
            return f"Failed to open {url}: {e}"

    async def search_web(self, query: str, engine: str = "google") -> str:
        """Search the web using Google or DuckDuckGo."""
        query = query.strip()
        encoded = urllib.parse.quote(query)
        if engine.lower() == "duckduckgo":
            url = f"https://duckduckgo.com/?q={encoded}"
        else:
            url = f"https://www.google.com/search?q={encoded}"
        return await self.open_url(url)

    async def open_youtube(self, query: str) -> str:
        """Search YouTube for a video/topic and immediately open and play the top result."""
        query = query.strip()
        # If the user passed an actual YouTube URL directly
        if "youtube.com" in query or "youtu.be" in query:
            await self.open_url(query)
            return "Opened YouTube link."

        # Fetch top video ID and title asynchronously with fast fallback
        loop = asyncio.get_running_loop()
        try:
            video_info = await asyncio.wait_for(
                loop.run_in_executor(None, self._resolve_youtube_video, query),
                timeout=1.4,
            )
        except asyncio.TimeoutError:
            video_info = None

        if video_info:
            video_id, title = video_info
            watch_url = f"https://www.youtube.com/watch?v={video_id}"
            await self.open_url(watch_url)
            return f"Playing '{title}' on YouTube."
        else:
            # Fallback to YouTube search results page
            encoded = urllib.parse.quote(query)
            search_url = f"https://www.youtube.com/results?search_query={encoded}"
            await self.open_url(search_url)
            return f"Opened YouTube search for '{query}'."

    @staticmethod
    def _resolve_youtube_video(query: str) -> Optional[Tuple[str, str]]:
        """Scrapes the top video result ID and title from YouTube search HTML without external dependencies."""
        try:
            encoded = urllib.parse.quote(query)
            url = f"https://www.youtube.com/results?search_query={encoded}"
            req = urllib.request.Request(
                url,
                headers={"User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0"},
            )
            with urllib.request.urlopen(req, timeout=1.2) as resp:
                html = resp.read().decode("utf-8", errors="ignore")

            # Try to match videoId and title together
            match = re.search(
                r'\"videoRenderer\":\{\"videoId\":\"([a-zA-Z0-9_-]{11})\".*?\"title\":\{\"runs\":\[\{\"text\":\"(.*?)\"\}\]',
                html,
            )
            if match:
                video_id = match.group(1)
                title = match.group(2)
                return video_id, title

            # Fallback: match any videoId
            ids = re.findall(r'\"videoId\":\"([a-zA-Z0-9_-]{11})\"', html)
            if ids:
                return ids[0], query

            return None
        except Exception as e:
            log.warning(f"YouTube resolution failed: {e}")
            return None

File: src/tools/web.rs
use crate::core::error::{JarvisError, Result};
use crate::tools::Tool;
use async_trait::async_trait;
use regex::Regex;
use serde_json::{json, Value};
use std::path::PathBuf;
use std::process::Stdio;
use std::sync::Arc;
use std::time::Duration;
use tokio::process::Command;
use tokio::time::timeout;
use tracing::debug;

pub struct WebNavigator {
    browser_bin: PathBuf,
}

impl Default for WebNavigator {
    fn default() -> Self {
        Self::new()
    }
}

impl WebNavigator {
    pub fn new() -> Self {
        let browser_bin = which::which("omarchy-launch-browser")
            .or_else(|_| which::which("xdg-open"))
            .unwrap_or_else(|_| PathBuf::from("xdg-open"));
        Self { browser_bin }
    }

    pub async fn open_url(&self, url: &str) -> Result<String> {
        let trimmed = url.trim();
        let target_url = if !trimmed.starts_with("http://") && !trimmed.starts_with("https://") {
            format!("https://{trimmed}")
        } else {
            trimmed.to_string()
        };

        debug!("Opening URL in browser: {target_url}");
        let mut child = Command::new(&self.browser_bin)
            .arg(&target_url)
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()?;

        let _ = timeout(Duration::from_millis(200), child.wait()).await;
        Ok(format!("Opened {target_url} in browser."))
    }

    pub async fn search_web(&self, query: &str) -> Result<String> {
        let encoded = urlencoding::encode(query.trim());
        let url = format!("https://www.google.com/search?q={encoded}");
        self.open_url(&url).await
    }

    pub async fn open_youtube(&self, query: &str) -> Result<String> {
        let clean = query.trim();
        if clean.contains("youtube.com") || clean.contains("youtu.be") {
            return self.open_url(clean).await;
        }

        // Try to scrape video ID with quick timeout
        let resolve_res = timeout(
            Duration::from_millis(1500),
            Self::resolve_youtube_video(clean),
        )
        .await;

        match resolve_res {
            Ok(Some((video_id, title))) => {
                let watch_url = format!("https://www.youtube.com/watch?v={video_id}");
                self.open_url(&watch_url).await?;
                Ok(format!("Playing '{title}' on YouTube."))
            }
            _ => {
                // Fallback to YouTube search query
                let encoded = urlencoding::encode(clean);
                let search_url = format!("https://www.youtube.com/results?search_query={encoded}");
                self.open_url(&search_url).await?;
                Ok(format!("Opened YouTube search for '{clean}'."))
            }
        }
    }

    async fn resolve_youtube_video(query: &str) -> Option<(String, String)> {
        let encoded = urlencoding::encode(query);
        let url = format!("https://www.youtube.com/results?search_query={encoded}");

        let client = reqwest::Client::builder()
            .timeout(Duration::from_millis(1200))
            .user_agent("Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0")
            .build()
            .ok()?;

        let resp = client.get(&url).send().await.ok()?;
        let html = resp.text().await.ok()?;

        let re = Regex::new(
            r#""videoRenderer":\{"videoId":"([a-zA-Z0-9_-]{11})".*?"title":\{"runs":\[\{"text":"(.*?)"\}"#,
        )
        .ok()?;

        if let Some(caps) = re.captures(&html) {
            let id = caps.get(1).map(|m| m.as_str().to_string())?;
            let title = caps
                .get(2)
                .map(|m| m.as_str().to_string())
                .unwrap_or_else(|| query.to_string());
            return Some((id, title));
        }

        let id_re = Regex::new(r#""videoId":"([a-zA-Z0-9_-]{11})""#).ok()?;
        if let Some(caps) = id_re.captures(&html) {
            let id = caps.get(1).map(|m| m.as_str().to_string())?;
            return Some((id, query.to_string()));
        }

        None
    }
}

// -----------------------------------------------------------------------------
// Tool Implementations
// -----------------------------------------------------------------------------

pub struct OpenUrlTool {
    web: Arc<WebNavigator>,
}

impl OpenUrlTool {
    pub fn new(web: Arc<WebNavigator>) -> Self {
        Self { web }
    }
}

#[async_trait]
impl Tool for OpenUrlTool {
    fn name(&self) -> &'static str {
        "open_url"
    }

    fn description(&self) -> &'static str {
        "Open any website or URL in the default browser."
    }

    fn parameters_schema(&self) -> Value {
        json!({
            "type": "OBJECT",
            "properties": {
                "url": {
                    "type": "STRING",
                    "description": "Website URL to open (e.g. 'github.com', 'https://reddit.com')."
                }
            },
            "required": ["url"]
        })
    }

    async fn execute(&self, args: Value) -> Result<String> {
        let url = args["url"].as_str().ok_or_else(|| {
            JarvisError::ToolParameter("open_url".into(), "url string is required".into())
        })?;

        self.web.open_url(url).await
    }
}

pub struct SearchWebTool {
    web: Arc<WebNavigator>,
}

impl SearchWebTool {
    pub fn new(web: Arc<WebNavigator>) -> Self {
        Self { web }
    }
}

#[async_trait]
impl Tool for SearchWebTool {
    fn name(&self) -> &'static str {
        "search_web"
    }

    fn description(&self) -> &'static str {
        "Search Google for information or websites."
    }

    fn parameters_schema(&self) -> Value {
        json!({
            "type": "OBJECT",
            "properties": {
                "query": {
                    "type": "STRING",
                    "description": "The search query string."
                }
            },
            "required": ["query"]
        })
    }

    async fn execute(&self, args: Value) -> Result<String> {
        let query = args["query"].as_str().ok_or_else(|| {
            JarvisError::ToolParameter("search_web".into(), "query string is required".into())
        })?;

        self.web.search_web(query).await
    }
}

pub struct OpenYoutubeTool {
    web: Arc<WebNavigator>,
}

impl OpenYoutubeTool {
    pub fn new(web: Arc<WebNavigator>) -> Self {
        Self { web }
    }
}

#[async_trait]
impl Tool for OpenYoutubeTool {
    fn name(&self) -> &'static str {
        "open_youtube"
    }

    fn description(&self) -> &'static str {
        "Search and play a video on YouTube (e.g. 'mkbhd latest video', 'lofi beats')."
    }

    fn parameters_schema(&self) -> Value {
        json!({
            "type": "OBJECT",
            "properties": {
                "query": {
                    "type": "STRING",
                    "description": "Search term or YouTube query."
                }
            },
            "required": ["query"]
        })
    }

    async fn execute(&self, args: Value) -> Result<String> {
        let query = args["query"].as_str().ok_or_else(|| {
            JarvisError::ToolParameter("open_youtube".into(), "query string is required".into())
        })?;

        self.web.open_youtube(query).await
    }
}
